proposals/dom_execute.md

```diff
@@ -0,0 +1,174 @@
+# Proposal: browser.dom.execute()
+
+**Summary**
+
+This API allows an isolated world (such as a content script world or user
+script world) to synchronously execute code in the main world of the document
+its injected in.
+
+**Document Metadata**
+
+**Author:** rdcronin
+
+**Sponsoring Browser:** Chromee
+
+**Contributors:** Rob--W, ...
+
+**Created:** 2024-06-24
+
+**Related Issues:** <TODO>
+
+## Motivation
+
+### Objective
+
+This allows isolated worlds to synchronously inject in the main world.  This is
+both beneficial by itself and also a prerequisite for our current plan to
+implement inter-world communication (the latter of which will be a separate
+proposal).
+
+#### Use Cases
+
+##### Executing script in the main world
+
+Today, extensions can execute script in the main world by:
+* Leveraging scripting.executeScript() or registering content scripts or user
+  scripts, or
+* Appending a new script tag to the document
+
+The former is heavily asynchronous or requires knowledge of whether to inject
+beforehand (to register a script).  The latter is very visible to the site and
+has more potential side effects.  While there's no way to fully "hide" a script
+from a site, it is desirable to avoid adding DOM elements (which could
+interfere with sites with certain properties).
+
+### Known Consumers
+
+User script managers have expressed an interest in this API, but this is also
+useful to any extension that wishes to execute script in the main world while
+leveraging information from the isolated world.
+
+## Specification
+
+### Schema
+
+```
+declare namespace dom {
+  interface ScriptInjection {
+    func: () => any;
+    args: any[];
+  };
+
+  export function execute(
+      injection: ScriptInjection
+  ): any
+}
+```
+
+### Behavior
+
+#### Function execution
+
+The function is executed _synchronously_ in the main world.  The return value
+is the serialized return value from the function execution.
+
+#### Cross-world contamination
+
+To prevent any "piercing" of the different worlds, all pieces are serialized in
+their original world, then a copy is deserialized in the target world.
+Otherwise, there would be a risk of sharing variables between worlds (even when
+non-obvious, such as by accessing the prototype of an item).
+
+##### Argument serialization
+
+Arguments are serialized and deserialized using the Structured Cloning
+algorithm.  This allows for more flexibility than simply JSON conversion.  In
+the future, some arguments may have custom serialization.
```

Comment:
Have the proposal authors considered whether arguments should support [transferable objects](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Transferable_objects)?

Reply:
We did consider it, but opted for structured cloning in order to reduce the implementation cost (allowing us to get this out sooner).  I also suspect transferables may have some interesting considerations with cross-world boundaries.  I've added this as potential future work.